File: java/core/src/main/scala/io/pdal/PointView.scala
/******************************************************************************
  * Copyright (c) 2016, hobu Inc.  (info@hobu.co)
  *
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following
  * conditions are met:
  *
  *     * Redistributions of source code must retain the above copyright
  *       notice, this list of conditions and the following disclaimer.
  *     * Redistributions in binary form must reproduce the above copyright
  *       notice, this list of conditions and the following disclaimer in
  *       the documentation and/or other materials provided
  *       with the distribution.
  *     * Neither the name of Hobu, Inc. nor the names of its
  *       contributors may be used to endorse or promote products derived
  *       from this software without specific prior written permission.
  *
  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
  * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
  * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
  * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
  * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
  * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
  * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
  * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
  * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
  * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
  * OF SUCH DAMAGE.
  ****************************************************************************/

package io.pdal

import java.nio.{ByteBuffer, ByteOrder}

class PointView extends Native {
  def getPointCloud(idx: Long): PointCloud = getPointCloud(idx, layout.dimTypes())
  def getPointCloud(idx: Long, dims: Array[DimType]): PointCloud =
    PointCloud(
      bytes    = getPackedPoint(idx, dims),
      dimTypes = layout.toSizedDimTypes(dims)
    )

  def getPointCloud(): PointCloud = getPointCloud(layout.dimTypes())
  def getPointCloud(dims: Array[DimType]): PointCloud =
    PointCloud(
      bytes    = getPackedPoints(dims),
      dimTypes = layout.toSizedDimTypes(dims)
    )

  def getPackedPoint(idx: Long): Array[Byte] = getPackedPoint(idx, layout.dimTypes())
  def getPackedPoints(): Array[Byte] = getPackedPoints(layout.dimTypes())
  def findDimType(name: String): DimType = layout.findDimType(name)
  def length(): Int = size()
  def getCrsWKT(): String = getCrsWKT(pretty = false)

  /**
    * Reads a packed point by point id from a set of packed points.
    */
  def get(idx: Int, packedPoints: Array[Byte]): Array[Byte] = get(idx, packedPoints, layout.dimTypes())
  def get(idx: Int, packedPoints: Array[Byte], dims: Array[DimType]): Array[Byte] = {
    val pointSize = dims.map(layout.dimSize(_)).sum.toInt
    val from = idx * pointSize
    val result = new Array[Byte](pointSize)
    var j = 0
    while(j < pointSize) {
      result(j) = packedPoints(from + j)
      j += 1
    }
    result
  }

  /**
    * Reads dim from a packed point, point should contain all layout dims.
    */
  def get(packedPoint: Array[Byte], dim: DimType): ByteBuffer = {
    val from = layout.dimPackedOffset(dim).toInt
    val dimSize = layout.dimSize(dim).toInt
    val result = new Array[Byte](dimSize)
    var j = 0
    while(j < dimSize) {
      result(j) = packedPoint(from + j)
      j += 1
    }
    ByteBuffer.wrap(result).order(ByteOrder.nativeOrder())
  }

  def getDouble(packedPoint: Array[Byte], dim: String): Double = getDouble(packedPoint, findDimType(dim))
  def getDouble(packedPoint: Array[Byte], dim: DimType): Double = get(packedPoint, dim).getDouble

  def getFloat(packedPoint: Array[Byte], dim: String): Float = getFloat(packedPoint, findDimType(dim))
  def getFloat(packedPoint: Array[Byte], dim: DimType): Float = get(packedPoint, dim).getFloat

  def getLong(packedPoint: Array[Byte], dim: String): Long = getLong(packedPoint, findDimType(dim))
  def getLong(packedPoint: Array[Byte], dim: DimType): Long = get(packedPoint, dim).getLong

  def getInt(packedPoint: Array[Byte], dim: String): Int = getInt(packedPoint, findDimType(dim))
  def getInt(packedPoint: Array[Byte], dim: DimType): Int = get(packedPoint, dim).getInt

  def getShort(packedPoint: Array[Byte], dim: String): Short = getShort(packedPoint, findDimType(dim))
  def getShort(packedPoint: Array[Byte], dim: DimType): Short = get(packedPoint, dim).getShort

  def getChar(packedPoint: Array[Byte], dim: String): Char = getChar(packedPoint, findDimType(dim))
  def getChar(packedPoint: Array[Byte], dim: DimType): Char = get(packedPoint, dim).getChar

  def getByte(packedPoint: Array[Byte], dim: String): Byte = getByte(packedPoint, findDimType(dim))
  def getByte(packedPoint: Array[Byte], dim: DimType): Byte = get(packedPoint, dim).get()

  /**
    * One dimension read; for multiple dims custom logic required.
    */

  def getDouble(idx: Int, dim: String): Double = getDouble(idx, findDimType(dim))
  def getDouble(idx: Int, dim: DimType): Double = get(idx, dim).getDouble

  def getFloat(idx: Int, dim: String): Float = getFloat(idx, findDimType(dim))
  def getFloat(idx: Int, dim: DimType): Float = get(idx, dim).getFloat

  def getLong(idx: Int, dim: String): Long = getLong(idx, findDimType(dim))
  def getLong(idx: Int, dim: DimType): Long = get(idx, dim).getLong

  def getInt(idx: Int, dim: String): Int = getInt(idx, findDimType(dim))
  def getInt(idx: Int, dim: DimType): Int = get(idx, dim).getInt

  def getShort(idx: Int, dim: String): Short = getShort(idx, findDimType(dim))
  def getShort(idx: Int, dim: DimType): Short = get(idx, dim).getShort

  def getChar(idx: Int, dim: String): Char = getChar(idx, findDimType(dim))
  def getChar(idx: Int, dim: DimType): Char = get(idx, dim).getChar

  def getByte(idx: Int, dim: String): Byte = getByte(idx, findDimType(dim))
  def getByte(idx: Int, dim: DimType): Byte = get(idx, dim).get()

  def get(idx: Int, dim: String): ByteBuffer = get(idx, findDimType(dim))
  def get(idx: Int, dim: DimType): ByteBuffer =
    ByteBuffer.wrap(getPackedPoint(idx, Array(dim))).order(ByteOrder.nativeOrder())

  def getX(idx: Int): Double = getDouble(idx, DimType.X)
  def getY(idx: Int): Double = getDouble(idx, DimType.Y)
  def getZ(idx: Int): Double = getDouble(idx, DimType.Z)

  def getX(packedPoint: Array[Byte]): Double = getDouble(packedPoint, DimType.X)
  def getY(packedPoint: Array[Byte]): Double = getDouble(packedPoint, DimType.Y)
  def getZ(packedPoint: Array[Byte]): Double = getDouble(packedPoint, DimType.Z)

  @native def layout(): PointLayout
  @native def size(): Int
  @native def empty(): Boolean
  @native def getCrsProj4(): String
  @native def getCrsWKT(pretty: Boolean): String
  @native def getPackedPoint(idx: Long, dims: Array[DimType]): Array[Byte]
  @native def getPackedPoints(dims: Array[DimType]): Array[Byte]
  @native def dispose(): Unit
}

File: pdal/PointView.cpp
/******************************************************************************
* Copyright (c) 2014, Hobu Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following
* conditions are met:
*
*     * Redistributions of source code must retain the above copyright
*       notice, this list of conditions and the following disclaimer.
*     * Redistributions in binary form must reproduce the above copyright
*       notice, this list of conditions and the following disclaimer in
*       the documentation and/or other materials provided
*       with the distribution.
*     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
*       names of its contributors may be used to endorse or promote
*       products derived from this software without specific prior
*       written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
* OF SUCH DAMAGE.
****************************************************************************/

#include <iomanip>

#include <pdal/KDIndex.hpp>
#include <pdal/PointView.hpp>
#include <pdal/util/Algorithm.hpp>

#include "private/Raster.hpp"

namespace pdal
{

int PointView::m_lastId = 0;

PointView::PointView(PointTableRef pointTable) : m_pointTable(pointTable),
    m_layout(pointTable.layout()), m_size(0), m_id(0)
{
	m_id = ++m_lastId;
}

PointView::PointView(PointTableRef pointTable, const SpatialReference& srs) :
	m_pointTable(pointTable), m_layout(pointTable.layout()), m_size(0),
    m_id(0), m_spatialReference(srs)
{
	m_id = ++m_lastId;
}

PointView::~PointView()
{}


PointViewIter PointView::begin()
{
    return PointViewIter(this, 0);
}


PointViewIter PointView::end()
{
    return PointViewIter(this, size());
}


PointId PointView::tableId(PointId idx)
{
    if (idx > size())
        throw pdal_error("Point index must increment.");
    if (idx == size())
    {
        PointId rawId = m_pointTable.addPoint();
        m_index.push_back(rawId);
        m_size++;
        assert(m_temps.empty());
        return rawId;
    }
    return m_index[idx];
}


void PointView::setFieldInternal(Dimension::Id dim, PointId idx,
    const void *buf)
{
    PointId rawId = 0;
    if (idx == size())
    {
        rawId = m_pointTable.addPoint();
        m_index.push_back(rawId);
        m_size++;
        assert(m_temps.empty());
    }
    else if (idx > size())
    {
        std::cerr << "Point index must increment.\n";
        //error - throw?
        return;
    }
    else
    {
        rawId = m_index[idx];
    }
    m_pointTable.setFieldInternal(dim, rawId, buf);
}


void PointView::calculateBounds(BOX2D& output) const
{
    for (PointId idx = 0; idx < size(); idx++)
    {
        double x = getFieldAs<double>(Dimension::Id::X, idx);
        double y = getFieldAs<double>(Dimension::Id::Y, idx);

        output.grow(x, y);
    }
}


void PointView::calculateBounds(BOX3D& output) const
{
    for (PointId idx = 0; idx < size(); idx++)
    {
        double x = getFieldAs<double>(Dimension::Id::X, idx);
        double y = getFieldAs<double>(Dimension::Id::Y, idx);
        double z = getFieldAs<double>(Dimension::Id::Z, idx);

        output.grow(x, y, z);
    }
}


MetadataNode PointView::toMetadata() const
{
    MetadataNode node;

    const Dimension::IdList& dims = layout()->dims();

    for (PointId idx = 0; idx < size(); idx++)
    {
        MetadataNode pointnode = node.add(std::to_string(idx));
        for (auto di = dims.begin(); di != dims.end(); ++di)
        {
            double v = getFieldAs<double>(*di, idx);
            pointnode.add(layout()->dimName(*di), v);
        }
    }
    return node;
}


TriangularMesh *PointView::createMesh(const std::string& name)
{
    if (Utils::contains(m_meshes, name))
        return nullptr;
    auto res = m_meshes.insert(std::make_pair(name,
        std::unique_ptr<TriangularMesh>(new TriangularMesh)));
    if (res.second)
        return res.first->second.get();
    return nullptr;
}


TriangularMesh *PointView::mesh(const std::string& name)
{
    auto it = m_meshes.find(name);
    if (it != m_meshes.end())
        return it->second.get();
    if (name.empty() && m_meshes.size())
        return m_meshes.begin()->second.get();
    return nullptr;
}


Rasterd *PointView::createRaster(const std::string& name, const RasterLimits& limits)
{
    if (Utils::contains(m_rasters, name))
        return nullptr;
    auto res = m_rasters.insert(std::make_pair(name,
        std::unique_ptr<Rasterd>(new Rasterd(limits))));
    if (res.second)
        return res.first->second.get();
    return nullptr;
}


Rasterd *PointView::raster(const std::string& name)
{
    auto it = m_rasters.find(name);
    if (it != m_rasters.end())
        return it->second.get();
    if (name.empty() && m_rasters.size())
        return m_rasters.begin()->second.get();
    return nullptr;
}


void PointView::invalidateProducts()
{
    m_index2.reset();
    m_index3.reset();
    // Should all meshes also be invalidated?
}


KD3Index& PointView::build3dIndex()
{
    //ABELL
    // Should we allow a force of point view build - perhaps the index has
    // changed or the point values have changed.
    if (!m_index3)
    {
        m_index3.reset(new KD3Index(*this));
        m_index3->build();
    }
    return *m_index3.get();
}


KD2Index& PointView::build2dIndex()
{
    //ABELL
    // Should we allow a force of point view build - perhaps the index has
    // changed or the point values have changed.
    if (!m_index2)
    {
        m_index2.reset(new KD2Index(*this));
        m_index2->build();
    }
    return *m_index2.get();
}


void PointView::dump(std::ostream& ostr) const
{
    using std::endl;
    PointLayoutPtr layout = m_pointTable.layout();
    const Dimension::IdList& dims = layout->dims();

    point_count_t numPoints = size();
    ostr << "Contains " << numPoints << "  points" << endl;
    for (PointId idx = 0; idx < numPoints; idx++)
    {
        ostr << "Point: " << idx << endl;

        for (auto di = dims.begin(); di != dims.end(); ++di)
        {
            Dimension::Id d = *di;
            const Dimension::Detail *dd = layout->dimDetail(d);
            ostr << layout->dimName(d) << " (" <<
                Dimension::interpretationName(dd->type()) << ") : ";

            switch (dd->type())
            {
            case Dimension::Type::Signed8:
                {
                    ostr << (int)(getFieldInternal<int8_t>(d, idx));
                    break;
                }
            case Dimension::Type::Signed16:
                {
                    ostr << getFieldInternal<int16_t>(d, idx);
                    break;
                }
            case Dimension::Type::Signed32:
                {
                    ostr << getFieldInternal<int32_t>(d, idx);
                    break;
                }
            case Dimension::Type::Signed64:
                {
                    ostr << getFieldInternal<int64_t>(d, idx);
                    break;
                }
            case Dimension::Type::Unsigned8:
                {
                    ostr << (unsigned)(getFieldInternal<uint8_t>(d, idx));
                    break;
                }
            case Dimension::Type::Unsigned16:
                {
                    ostr << getFieldInternal<uint16_t>(d, idx);
                    break;
                }
            case Dimension::Type::Unsigned32:
                {
                    ostr << getFieldInternal<uint32_t>(d, idx);
                    break;
                }
            case Dimension::Type::Unsigned64:
                {
                    ostr << getFieldInternal<uint64_t>(d, idx);
                    break;
                }
            case Dimension::Type::Float:
                {
                    ostr << getFieldInternal<float>(d, idx);
                    break;
                }
            case Dimension::Type::Double:
                {
                    ostr << getFieldInternal<double>(d, idx);
                    break;
                }
            case Dimension::Type::None:
                ostr << "NONE";
                break;
            }
            ostr << endl;
        }
    }
}


std::ostream& operator<<(std::ostream& ostr, const PointView& buf)
{
    buf.dump(ostr);
    return ostr;
}

} // namespace pdal
